nsiqcppstyle_rulemanager.py

```diff
@@ -172,77 +192,93 @@ def RunProjectRules(self, targetName):
+        self.functionScopeRules.append(user_function)
 
-    def AddFunctionNameRule(self, functionRule):
-        """ Add rule on the function name place"""
-        self.functionNameRules.append(functionRule)
+    def AddFunctionNameRule(self,
+                            user_function: Callable[
+                                [LexerType, FullNameType, DeclarationType, ContextStackType, ContextType],
+                                None]):
+        """ Add rule on the function name place """
+        self.functionNameRules.append(user_function)
 
-    def AddLineRule(self, lineRule):
+    def AddLineRule(self, user_function: Callable[[LexerType, LineType, LineNumberType], None]):
         """ Add rule on the each line """
-        self.lineRules.append(lineRule)
+        self.lineRules.append(user_function)
 
-    def AddRule(self, rule):
+    def AddRule(self, user_function: Callable[[LexerType, ContextStackType], None]):
         """ Add rule on any token """
-        self.rules.append(rule)
+        self.rules.append(user_function)
 
-    def AddTypeNameRule(self, typeNameRule):
-        """ Add rule on any type (class / struct / union / namesapce / enum) """
-        self.typeNameRules.append(typeNameRule)
+    def AddTypeNameRule(self, user_function: Callable[
+                                [LexerType, TypeNameType, TypeFullNameType,
+                                 DeclarationType, ContextStackType, TypeContextType],
+                                None]):
+        """ Add rule on any type (class / struct / union / namespace / enum) """
+        self.typeNameRules.append(user_function)
 
-    def AddTypeScopeRule(self, typeScopeRule):
-        """ Add rule on the any type definition scope """
-        self.typeScopeRules.append(typeScopeRule)
+    def AddTypeScopeRule(self, user_function: Callable[[LexerType, ContextStackType], None]):
+        """ Add rule when the token is within a type definition scope """
+        self.typeScopeRules.append(user_function)
 
-    def AddFileEndRule(self, fileEndRule):
-        """
-        Add rule on the file end
-        Added Rule should be function with following prototype "def RunRule(lexer, filename, dirname)"
-        lexer is the lexer used to analyze the source. it points the end token of source.
-        filename is the filename analyzed.
-        dirname is the file directory.
-        """
-        self.fileEndRules.append(fileEndRule)
+    def AddFileEndRule(self, user_function: Callable[[LexerType, FileNameType, DirNameType], None]):
+        """ Add rule on the file end """
+        self.fileEndRules.append(user_function)
+
+    def AddFileStartRule(self, user_function: Callable[[LexerType, FileNameType, DirNameType], None]):
+        """ Add rule on the file start """
+        self.fileStartRules.append(user_function)
 
-    def AddFileStartRule(self, fileStartRule):
+    def AddSessionEndRule(self, user_function: Callable[[], None]):

```

Comment:
Should we state that we expect a `None` result, or that the result can be `Any` but we wouldn't use it?

The former is more restrictive on the user (lambda functions can't be used since they always return), but the latter gives a false direction to the user that some return value is expected.

Thoughts?

Reply:
Personally I prefer to be truthful :-) and use the "-> None" notation.  I don't see someone using the "Add*" functions within a lambda, do you?

Reply:
Let's keep the None